Reapply "[analyzer] Accept C library functions from the `std` namespace" again

NagyDonat opened the pull request:
This reapplies 80ab8234ac309418637488b97e0a62d8377b2ecf again, after fixing a name collision warning in the unit tests (see the revert commit 13ccaf9b9d4400bb128b35ff4ac733e4afc3ad1c for details).

NagyDonat committed:
Reapply "[analyzer] Accept C library functions from the `std` namespace" again

This reapplies 80ab8234ac309418637488b97e0a62d8377b2ecf again, after
fixing a name collision warning in the unit tests (see the revert commit
13ccaf9b9d4400bb128b35ff4ac733e4afc3ad1c for details).

Co-authored-by: Balazs Benics <benicsbalazs@gmail.com>

NagyDonat committed:
Satisfy git-clang-format

Szelethus reviewed (approved):
Lets hope it works fine this time around :)

NagyDonat commented:
I'll delay reapplying this commit a bit because I realized that it has minor unintended side effects in MallocChecker.

NagyDonat committed:
Fix recognition of POSIX getline in MallocChecker

The name "getline" is used for two completely different functions:
`std::getline` from the C++ standard library works with streams and
strings, while the C function `getline` defined by POSIX and the dynamic
memory TS works with C data structures and `malloc` style allocations.

This commit adds a manual `isInStdNamespace()` check to discard calls to
`std::getline` in MallocChecker. In theory we could keep a separate
matching mode for "only a C function, not from the std namespace" but I
think that situations like this `getline` are so rare that it's better
to handle them individually.

Note that the analyzer was apparently producing correct behavior even
without this fix, because the unexpected `std::getline` calls were
silently discarded (as their arguments had unexpected types and the
callback just bailed out).

NagyDonat commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -446,8 +448,11 @@ class MallocChecker
-      {{{"getdelim"}, 4}, &MallocChecker::checkGetdelim},
+
+      // NOTE: the following CallDescription also matches the C++ standard
+      // library function std::getdelim(); the callback will filter it out.
```
```suggestion
      // library function std::getline(); the callback will filter it out.
```

NagyDonat committed:
Fix a mistake in comments

NagyDonat committed:
Merge branch 'main' into CallDescription_fix_unittest_again

NagyDonat commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -1435,9 +1440,17 @@ void MallocChecker::checkGMallocN0(const CallEvent &Call,
+static bool isFromStdNamespace(const CallEvent &Call) {
+  const Decl *FD = Call.getDecl();
+  assert(FD && "a CallDescription cannot match a call without a Decl");
+  return (FD->isInStdNamespace());
```
```suggestion
  return FD->isInStdNamespace();
```

NagyDonat committed:
Remove superfluous parentheses

steakhal commented:
This fixed tickets in our backlog. Thanks for pushing for this.
